FiltersEngine.cpp

```diff
@@ -0,0 +1,201 @@
+        _nextBitPosition++;
+    }
+
+    if (_outputs.count(outputName) == 0) {
+        _outputs.insert(outputName);
+    }
+
+    return ret;
+}
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::AddFilterList(std::vector<ProcFilterSpec>& pfsVec, std::string& outputName)
+{
+    std::bitset<FILTER_BITSET_SIZE> ret;
+
+    for (auto pfs : pfsVec) {
+        ret |= AddFilter(pfs, outputName);
+    }
+
+    return ret;
+}
+
+bool FiltersEngine::ProcessMatchFilter(std::shared_ptr<ProcessTreeItem> process, ProcFilterSpec& pfs, unsigned int height)
+{
+    if (pfs._depth != -1 && pfs._depth < height) {
+        return false;
+    }
+
+    if (pfs._match_mask & PFS_MATCH_UID) {
+        if (pfs._uid != process->_uid) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_GID) {
+        if (pfs._gid != process->_gid) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_EQUALS) {
+        if (pfs._exeMatchValue != process->_exe) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_STARTSWITH) {
+        if (process->_exe.substr(0, pfs._exeMatchValue.length()) != pfs._exeMatchValue) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_CONTAINS) {
+        if (process->_exe.find(pfs._exeMatchValue) == std::string::npos) {
+            return false;
+        }
+    }
+    if (pfs._match_mask & PFS_MATCH_EXE_REGEX) {
+        if (!std::regex_search(process->_exe, pfs._exeRegex)) {
+            return false;
+        }
+    }
+
+    for (auto cf : pfs._cmdlineFilters) {
+        if (cf._matchType == MatchEquals) {
+            if (cf._matchValue != process->_cmdline) {
+                return false;
+            }
+        } else if (cf._matchType == MatchStartsWith) {
+            if (process->_cmdline.substr(0, cf._matchValue.length()) != cf._matchValue) {
+                return false;
+            }
+        } else if (cf._matchType == MatchContains) {
+            if (process->_cmdline.find(cf._matchValue) == std::string::npos) {
+                return false;
+            }
+        } else if (cf._matchType == MatchRegex) {
+            if (!std::regex_search(process->_cmdline, cf._matchRegex)) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::GetFlags(std::shared_ptr<ProcessTreeItem> process, unsigned int height)
+{
+    std::bitset<FILTER_BITSET_SIZE> flags;
+
+    for (auto element : _filtersBitPosition) {
+        ProcFilterSpec pfs = element.first;
+        if (ProcessMatchFilter(process, pfs, height)) {
+            flags[element.second.bitPosition] = 1;
+        }
+    }
+
+    return flags;
+}
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::GetCommonFlagMask()
+{
+    std::bitset<FILTER_BITSET_SIZE> flags;
+    unsigned int numberOfOutputs = _outputs.size();
+
+    for (auto element : _filtersBitPosition) {
```

Comment:
Instead of re-iterating over _filtersBitPosition each time, consider caching the value and only re-calculating it if a new filter is added.

Reply:
Accepted.